docs/spec/governance/software_upgrade.md

```diff
@@ -0,0 +1,133 @@
+## Software Upgrade
+
+Software is upgraded according to a 4-step process.
+
+1. Agree on software upgrade plan via governance (currently via a plaintext proposal).
+2. Once the upgrade plan is accepted, validators each update their software according to proposal from step 1.
+3. Each upgraded validator submits a transaction to the SoftwareUpgradeKeeper.
+4. Once a sufficient quorum of validators has completed step 3, new logic kicks
+   in in the form of a special parameter.
+
+### Step 1: Agree on the Upgrade Plan
+
+First, an English plaintext proposal should be submitted based on the following template:
+
+```
+# SOFTWARE UPGRADE PROPOSAL
+
+This is a proposal to upgrade the software.
+
+## Outline:
+
+* Current version: [XXX]
+* Current version commit hash: [XXX]
+* New version Github URL: https://github.com/...
+* New version: [XXX]
+* New version commit hash: [XXX]
+* Upgrade prepared by:
+ - <name/handle> on <date>; commit signature: https://github.com/...
+* Audited by:
+ - <name/handle> on <date>; report: https://github.com/...
+ - <name/handle> on <date>; report: https://github.com/...
+
+## Description:
+
+Describe the purpose of the upgrade and other details.
+
+- [ ] Link to any previous discussions
+- [ ] Link to any alternative proposals
+
+## Requirements:
+
+Describe any requirements
+
+- [ ] Any new hardware requirements?
+- [ ] Any new bandwidth requirements?
+- [ ] Any dependencies on other proposals?
+
+## Procedure:
+
+* Deadline: By 2 weeks from after the proposal has been accepted.
+
+Once this proposal has been accepted, validators should upgrade to the new
+software as described above by the deadline.  Once the new software is
+deployed, the new software will automaticaly submit a message to the
+SoftwareUpgradeKeeper.  The SoftwareUpgradeKeeper ensures that a sufficient
+quorum of validators (SOFTWARE_UPGRADE_QUORUM parameter) is running a specific
+version of the blockchain software.
+
+Once a sufficient quorum has been reached, the "SOFTWARE_UPGRADE_[XXX]_ENABLED"
+parameter will get set to true, thus switching on the new logic.
+
+## Conflict Resolution
+
+In case there are conflicting proposals, the most recent proposal to get
```

Comment:
To get accepted or to pass the `SoftwareUpgradeKeeper` quorum requirement? What if two proposals are accepted but the latter passes the quorum first?

Reply:
We could have a second accepted proposal reset the `SoftwareUpgradeKeeper`, useful if a proposal is passed and validators start upgrading but discover an issue.

Reply:
Oh! This gives me an idea.

Maybe we don't need a second proposal, but just the ability for individual validators to reset/cancel their entry in the keeper's map. A proposal won't pass if the quorum is not met. The only problem there is slashing, but we could have the slashing condition require a minimum quorum. Like if less than a third of validators signal, noone gets slashed. 

Only if `1/3 < number of validators that signalled < Switch-triggering quorum` and deadline is met do we slash validators that did not signalled.